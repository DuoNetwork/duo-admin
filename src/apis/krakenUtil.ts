import mysqlUtil from '../mysqlUtil';
import * as CST from '../constants';
const Kraken = require('kraken-wrapper');

// let dbConn;

const INTERVAL_SECS = 2;

let last = 0; // last = id to be used as since when polling for new trade data
let requestJson: object = {};

export class KrakenUtil {
	async fetchETHTradesByOwnWebSocket() {
		const kraken = new Kraken();

		if (last == 0) {
			requestJson = { pair: 'ETHUSD' };
		} else if (last != undefined) {
			requestJson = { pair: 'ETHUSD', last: last };
		}
		console.log('request: ' + last + 'length: ' + last.toString().split('.')[0].length);

		try {
			const response = await kraken.getTrades(requestJson);
			// var jsonObj= JSON.parse(response);

			const returnFirstLevelArray = response.result.XETHZUSD;
			// console.log(returnFirstLevelArray);

			returnFirstLevelArray.forEach(secondLevelArr => {
				let trade_type: string = 'buy';
				const exchange_returned_timestamp =
					Math.floor(Number(secondLevelArr[2]) * 1000) + '';

				if (secondLevelArr[3] == 'b') {
					trade_type = 'buy';
				} else if (secondLevelArr[3] == 's') {
					trade_type = 'sell';
				}
				mysqlUtil.insertDataIntoMysql(
					CST.EXCHANGE_KRAKEN,
					'',
					secondLevelArr[0],
					secondLevelArr[1],
					trade_type,
					exchange_returned_timestamp
				);
			});

			last = response.result.last;
			console.log(last);
		} catch (error) {
			console.log(error);
		}
	}

	startFetching() {
		setInterval(this.fetchETHTradesByOwnWebSocket, INTERVAL_SECS * 1000);
	}
}
const krakenUtil = new KrakenUtil();
export default krakenUtil;
