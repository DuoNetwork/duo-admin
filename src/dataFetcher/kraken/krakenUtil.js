'use strict';

var MysqlUtil = require('../Utils/sharedMysql');

var dbConn;

const INTERVAL_SECS = 2;

const EXCHANGE_NAME = 'KRANKEN';
const db_host = 'localhost';
const db_user = 'root';
const db_password = '123456';
const db_name = 'priceFeedDB';
const db_table_name = 'ETH_Trades_Table';

var last = 0; // last = id to be used as since when polling for new trade data
var requestJson = '';

class KrankenTradeFeedUtil {
	initDB() {
		console.log('Init the DB');

		mysqlUtil.setup(EXCHANGE_NAME, db_host, db_user, db_password, db_name, db_table_name);

		mysqlUtil.initDB();
	}

	fetchETHTradesByOwnWebSocket() {
		const Kraken = require('kraken-wrapper');

		const kraken = new Kraken();

		if (last == 0) {
			requestJson = { pair: 'ETHUSD' };
		} else if (last != undefined) {
			requestJson = { pair: 'ETHUSD', last: last };
		}
		console.log('request: ' + last + 'length: ' + last.toString().split('.')[0].length);

		kraken
			.getTrades(requestJson)
			.then(response => {
				// var jsonObj= JSON.parse(response);

				dbConn = mysqlUtil.dbConn;

				if (dbConn == undefined) {
					krankenTradeFeedUtil.initDB();
				}

				var returnFirstLevelArray = response.result.XETHZUSD;

				returnFirstLevelArray.forEach(function(secondLevelArr) {
					var trade_type = 'buy';

					if (secondLevelArr[3] == 'b') {
						trade_type = 'buy';
					} else if (secondLevelArr[3] == 's') {
						trade_type = 'sell';
					}
					mysqlUtil.insertDataIntoMysql(EXCHANGE_NAME, '', secondLevelArr[0], secondLevelArr[1], trade_type, secondLevelArr[2]);
				});

				last = response.result.last;
				console.log(last);
			})
			.catch(error => {
				console.log(error);
			});
	}

	startFetching() {
		setInterval(this.fetchETHTradesByOwnWebSocket, INTERVAL_SECS * 1000);
	}
}
let mysqlUtil = new MysqlUtil();
let krankenTradeFeedUtil = new KrankenTradeFeedUtil();
// krankenTradeFeedUtil.fetchETHTradesByOwnWebSocket();
export default krankenTradeFeedUtil;

/*
    https://www.kraken.com/help/api#get-recent-trades

Input:

pair = asset pair to get trade data for
since = return trade data since given id (optional.  exclusive)

Output

<pair_name> = pair name
    array of array entries(<time>, <bid>, <ask>)
last = id to be used as since when polling for new spread data


Kraken实际的return value：
    array of array entries(<time>, <bid>, <ask>)

    price,       amount,         timestamp       trade type
[ '556.70000', '0.81900000', 1524154859.9084, 's', 'l', '' ]
[ '557.62000', '0.30160936', 1524154843.7233, 'b', 'l', '' ]

 */
