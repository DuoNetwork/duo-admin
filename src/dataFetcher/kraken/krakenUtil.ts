import MysqlUtil from '../../utils/MysqlUtil';
import * as CST from '../../constant';

// let dbConn;

const INTERVAL_SECS = 2;

const EXCHANGE_NAME = CST.EXCHANGE_KRAKEN;
const DB_HOST = CST.DB_HOST;
const DB_USER = CST.DB_USER;
const DB_PASSWORD = CST.DB_PASSWORD;
const DB_PRICEFEED = CST.DB_PRICEFEED;
const DB_TABLE_TRADE = CST.DB_TABLE_TRADE;

var last = 0; // last = id to be used as since when polling for new trade data
var requestJson: object = {};

export class KrankenTradeFeedUtil {
	mysqlUtil: MysqlUtil;

	constructor() {
		console.log('begin');
		this.mysqlUtil = new MysqlUtil(EXCHANGE_NAME, DB_HOST, DB_USER, DB_PASSWORD, DB_PRICEFEED, DB_TABLE_TRADE);
	}

	initDB() {
		console.log('Init the DB');


		this.mysqlUtil.initDB();
	}

	fetchETHTradesByOwnWebSocket() {
		const Kraken = require('kraken-wrapper');

		const kraken = new Kraken();

		if (last == 0) {
			requestJson = { pair: 'ETHUSD' };
		} else if (last != undefined) {
			requestJson = { pair: 'ETHUSD', last: last };
		}
		console.log('request: ' + last + 'length: ' + last.toString().split('.')[0].length);

		kraken
			.getTrades(requestJson)
			.then(response => {
				// var jsonObj= JSON.parse(response);

				let dbConn = krankenTradeFeedUtil.mysqlUtil.dbConn;

				if (dbConn == undefined) {
					krankenTradeFeedUtil.initDB();
				}

				var returnFirstLevelArray = response.result.XETHZUSD;

				returnFirstLevelArray.forEach((secondLevelArr) => {
					var trade_type = 'buy';

					if (secondLevelArr[3] == 'b') {
						trade_type = 'buy';
					} else if (secondLevelArr[3] == 's') {
						trade_type = 'sell';
					}
					krankenTradeFeedUtil.mysqlUtil.insertDataIntoMysql(EXCHANGE_NAME, '', secondLevelArr[0], secondLevelArr[1], trade_type, secondLevelArr[2]);
				});

				last = response.result.last;
				console.log(last);
			})
			.catch(error => {
				console.log(error);
			});
	}

	startFetching() {
		setInterval(krankenTradeFeedUtil.fetchETHTradesByOwnWebSocket, INTERVAL_SECS * 1000);
	}
}
const krankenTradeFeedUtil = new KrankenTradeFeedUtil();
export default krankenTradeFeedUtil;