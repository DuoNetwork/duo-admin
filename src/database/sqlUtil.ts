import * as mysql from 'mysql';
import * as CST from '../constants';
import { IPrice, ITrade } from '../types';
import util from '../util';
import dynamoUtil from './dynamoUtil';

export class SqlUtil {
	private conn: undefined | mysql.Connection = undefined;
	public init(host: string, user: string, pwd: string) {
		this.conn = mysql.createConnection({
			host: host,
			user: user,
			password: pwd,
			database: CST.DB_SQL_SCHEMA_PRICEFEED
		});

		this.conn.connect(err => {
			if (err) throw err;
			util.log('Connected!');
		});
	}

	public executeQuery(sqlQuery: string): Promise<any> {
		// util.log(sqlQuery);
		return new Promise((resolve, reject) => {
			if (this.conn)
				this.conn.query(sqlQuery, (err, result) => {
					if (err && err.code !== undefined && err.code === 'ER_DUP_ENTRY')
						// util.log('.');
						// rocess.stdout.write(".");
						reject(err);
					else if (err) reject(err);
					else resolve(result);
				});
			else reject('sql db connection is not initialized');
		});
	}

	public async insertTradeData(trade: ITrade, insertStatus: boolean) {
		const systemTimestamp = util.getNowTimestamp(); // record down the MTS

		// To do the checking for out of boundary data.
		// if(isNaN(trade.price)){
		// 	util.log('Price is NaN!');
		// 	return;
		// }

		// if(isNaN(trade.amount)){
		// 	util.log('Amount is NaN!');
		// 	return;
		// }

		// let price_str = math.format(price, { exponential: { lower: 1e-100, upper: 1e100 } });
		// let amount_str = math.format(amount, { exponential: { lower: 1e-100, upper: 1e100 } });

		const priceStr = trade.price.toString();
		const amountStr = trade.amount.toString();

		const sql =
			'REPLACE ' +
			CST.DB_SQL_TRADE +
			" VALUES ('" +
			trade.source +
			"','" +
			trade.id +
			"','" +
			priceStr +
			"','" +
			amountStr +
			"','" +
			(trade.timestamp || systemTimestamp) +
			"','" +
			systemTimestamp +
			"')";
		// util.log(await this.executeQuery(sql));
		await this.executeQuery(sql);
		if (insertStatus)
			await dynamoUtil.insertStatusData(
				dynamoUtil.convertTradeToSchema(trade, systemTimestamp)
			);
	}

	public async insertPrice(price: IPrice) {
		util.log(
			await this.executeQuery(
				'INSERT INTO ' +
					CST.DB_SQL_HISTORY +
					" VALUES ('" +
					price.timestamp +
					"','" +
					price.price +
					"','" +
					price.volume +
					"')"
			)
		);
		await dynamoUtil.insertStatusData(dynamoUtil.convertPriceToSchema(price));
	}

	public async readLastPrice(): Promise<IPrice> {
		const res = await this.executeQuery(
			'SELECT * FROM ' +
				CST.DB_SQL_HISTORY +
				' order by ' +
				CST.DB_HISTORY_TIMESTAMP +
				' DESC LIMIT 1'
		);
		return res[0]
			? {
					price: Number(res[0][CST.DB_HISTORY_PRICE]),
					timestamp: Number(res[0][CST.DB_HISTORY_TIMESTAMP]),
					volume: Number(res[0][CST.DB_HISTORY_VOLUME])
			}
			: { price: 0, timestamp: 0, volume: 0 };
	}

	public async readSourceData(currentTimestamp: number): Promise<ITrade[]> {
		const lowerTime = currentTimestamp - 3600000 + '';
		const upperTime = currentTimestamp + '';
		const res: object[] = await this.executeQuery(
			'SELECT * FROM ' +
				CST.DB_SQL_TRADE +
				' WHERE ' +
				CST.DB_TX_TS +
				' >= ' +
				lowerTime +
				' AND ' +
				CST.DB_TX_TS +
				' <= ' +
				upperTime
		);
		return res.map(item => ({
			source: item[CST.DB_TX_SRC],
			id: item[CST.DB_TX_ID],
			price: Number(item[CST.DB_TX_PRICE]),
			amount: Number(item[CST.DB_TX_AMOUNT]),
			timestamp: Number(item[CST.DB_TX_TS])
		}));
	}
}

const sqlUtil = new SqlUtil();
export default sqlUtil;
