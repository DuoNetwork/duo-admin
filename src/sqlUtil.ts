import * as mysql from 'mysql';
import * as CST from './constants';
// const math = require("mathjs");
import { Price, Trade } from './types';

export class SqlUtil {
	conn: undefined | mysql.Connection = undefined;

	init(user: string, pwd: string) {
		this.conn = mysql.createConnection({
			host: CST.DB_HOST,
			user: user,
			password: pwd,
			database: CST.DB_PRICEFEED
		});

		this.conn.connect(function(err) {
			if (err) throw err;
			console.log('Connected!');
		});
	}

	executeQuery(sqlQuery: string): any {
		return new Promise((resolve, reject) => {
			if (this.conn)
				this.conn.query(sqlQuery, (err, result) => {
					if (err && err.code != undefined && err.code === 'ER_DUP_ENTRY')
						// console.log('.');
						// rocess.stdout.write(".");
						reject(err);
					else if (err) reject(err);
					else resolve(result);
				});
			else reject('db connection is not initialized');
		});
	}

	async insertSourceData(sourceData: Trade) {
		const systemTimestamp = Math.floor(Date.now()); // record down the MTS

		// let price_str = math.format(price, { exponential: { lower: 1e-100, upper: 1e100 } });
		// let amount_str = math.format(amount, { exponential: { lower: 1e-100, upper: 1e100 } });

		const priceStr = sourceData.price.split('"').join('');
		const amountStr = sourceData.amount.split('"').join('');

		const sql =
			'REPLACE ' +
			CST.DB_TABLE_TRADE +
			" VALUES ('" +
			sourceData.source +
			"','" +
			sourceData.tradeId +
			"','" +
			priceStr +
			"','" +
			amountStr +
			"','" +
			sourceData.tradeType +
			"','" +
			(sourceData.sourceTimestamp || systemTimestamp + '') +
			"','" +
			systemTimestamp +
			"')";
		// console.log(await this.executeQuery(sql));
		await this.executeQuery(sql);
	}

	async insertPrice(price: Price) {
		console.log(
			await this.executeQuery(
				'INSERT INTO eth_historical_price' +
					" VALUES ('" +
					price.timestamp +
					"','" +
					price.price +
					"')"
			)
		);
	}

	async readLastPrice(): Promise<Price> {
		const res = await this.executeQuery(
			'SELECT * FROM eth_historical_price order by timestamp DESC LIMIT 1'
		);
		return {
			price: res[0][CST.DB_HISTORY_PRICE],
			timestamp: res[0][CST.DB_HISTORY_TIMESTAMP]
		};
	}

	async readSourceData(currentTimestamp: number): Promise<Trade[]> {
		const lowerTime = currentTimestamp - 3600000 + '';
		const upperTime = currentTimestamp + '';
		const res: object[] = await this.executeQuery(
			'SELECT * FROM ' +
				CST.DB_TABLE_TRADE +
				' WHERE exchange_returned_timestamp >= ' +
				lowerTime +
				' AND exchange_returned_timestamp <= ' +
				upperTime
		);
		return res.map(item => ({
			source: item[CST.DB_TX_EXCHANGE_SRC],
			tradeId: item[CST.DB_TX_TRADE_ID],
			price: item[CST.DB_TX_PRICE],
			amount: item[CST.DB_TX_AMOUNT],
			tradeType: item[CST.DB_TX_TYPE],
			sourceTimestamp: item[CST.DB_TX_EXCHANGE_TIME_STAMP]
		}));
	}
}

const sqlUtil = new SqlUtil();
export default sqlUtil;