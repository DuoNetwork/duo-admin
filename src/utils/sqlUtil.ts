import * as mysql from 'mysql';
import * as CST from '../common/constants';
import { IPriceFix, ITrade } from '../common/types';
import dynamoUtil from './dynamoUtil';
import util from './util';

class SqlUtil {
	private conn: undefined | mysql.Connection = undefined;
	public init(host: string, user: string, pwd: string) {
		this.conn = mysql.createConnection({
			host: host,
			user: user,
			password: pwd,
			database: CST.DB_SQL_SCHEMA_PRICEFEED
		});
		this.conn.on('error', err => {
			if (err.code === 'PROTOCOL_CONNECTION_LOST') {
				util.logInfo('ERROR: Server Disconnects. Reconnecting');
				this.init(host, user, pwd);
			} else throw err;
		});

		return new Promise<void>((resolve, reject) => {
			if (this.conn)
				this.conn.connect(err => {
					if (err) reject(err);
					util.logInfo('Connected!');
					resolve();
				});
			else reject('no connection');
		});
	}

	public executeQuery(sqlQuery: string): Promise<any> {
		// util.logInfo(sqlQuery);
		return new Promise((resolve, reject) => {
			if (this.conn)
				this.conn.query(sqlQuery, (err, result) => {
					if (err && err.code !== undefined && err.code === 'ER_DUP_ENTRY')
						// util.logInfo('.');
						// rocess.stdout.write(".");
						reject(err);
					else if (err) reject(err);
					else resolve(result);
				});
			else reject('sql db connection is not initialized');
		});
	}

	public getTradeCounts() {
		const sql = `SELECT COUNT(*) FROM ${CST.DB_SQL_TRADE} WHERE 1`;
		return this.executeQuery(sql);
	}

	public async insertTradeData(trade: ITrade, insertStatus: boolean) {
		const systemTimestamp = util.getUTCNowTimestamp(); // record down the MTS

		// To do the checking for out of boundary data.
		// if(isNaN(trade.price)){
		// 	util.logInfo('Price is NaN!');
		// 	return;
		// }

		// if(isNaN(trade.amount)){
		// 	util.logInfo('Amount is NaN!');
		// 	return;
		// }

		// let price_str = math.format(price, { exponential: { lower: 1e-100, upper: 1e100 } });
		// let amount_str = math.format(amount, { exponential: { lower: 1e-100, upper: 1e100 } });

		const priceStr = trade.price.toString();
		const amountStr = trade.amount.toString();

		const sql =
			'REPLACE ' +
			CST.DB_SQL_TRADE +
			" VALUES ('" +
			trade.source +
			"','" +
			trade.id +
			"','" +
			priceStr +
			"','" +
			amountStr +
			"','" +
			(trade.timestamp || systemTimestamp) +
			"','" +
			systemTimestamp +
			"','" +
			(trade.quote + '|' + trade.base) +
			"')";
		// util.logInfo(await this.executeQuery(sql));
		await this.executeQuery(sql);
		if (insertStatus)
			await dynamoUtil.insertStatusData(
				dynamoUtil.convertTradeToDynamo(trade, systemTimestamp)
			);
	}

	public async insertPrice(price: IPriceFix) {
		util.logInfo(
			await this.executeQuery(
				'INSERT INTO ' +
					CST.DB_SQL_HISTORY +
					" VALUES ('" +
					price.timestamp +
					"','" +
					price.price +
					"','" +
					price.volume +
					"','" +
					(price.quote + '|' + price.base) +
					"')"
			)
		);
	}

	public async readLastPrice(base: string, quote: string): Promise<IPriceFix> {
		const pair = quote + '|' + base;
		const res = await this.executeQuery(
			'SELECT * FROM ' +
				CST.DB_SQL_HISTORY +
				' order by ' +
				CST.DB_HISTORY_TIMESTAMP +
				` WHERE pair = '${pair}' DESC LIMIT 1`
		);
		return res[0]
			? {
					price: Number(res[0][CST.DB_HISTORY_PRICE]),
					timestamp: Number(res[0][CST.DB_HISTORY_TIMESTAMP]),
					volume: Number(res[0][CST.DB_HISTORY_VOLUME]),
					source: '',
					base: base,
					quote: quote
			}
			: {
					price: 0,
					timestamp: 0,
					volume: 0,
					source: '',
					base: base,
					quote: quote
			};
	}

	public async readSourceData(
		currentTimestamp: number,
		base: string,
		quote: string
	): Promise<ITrade[]> {
		const pair = quote + '|' + base;
		const lowerTime = currentTimestamp - 3600000 + '';
		const upperTime = currentTimestamp + '';
		const res: Array<{ [key: string]: string }> = await this.executeQuery(
			'SELECT * FROM ' +
				CST.DB_SQL_TRADE +
				' WHERE ' +
				CST.DB_TX_TS +
				' >= ' +
				lowerTime +
				' AND ' +
				CST.DB_TX_TS +
				' <= ' +
				upperTime +
				` AND pair = '${pair}';`
		);
		return res.map(item => ({
			quote: item[CST.DB_TX_QTE],
			base: item[CST.DB_TX_BASE],
			id: item[CST.DB_TX_ID],
			source: item[CST.DB_TX_SRC],
			price: Number(item[CST.DB_TX_PRICE]),
			amount: Number(item[CST.DB_TX_AMOUNT]),
			timestamp: Number(item[CST.DB_TX_TS])
		}));
	}

	public async cleanDB(): Promise<void> {
		const queryString =
			'DELETE FROM ' +
			CST.DB_SQL_TRADE +
			' WHERE timestamp < UNIX_TIMESTAMP(DATE_SUB(NOW(), INTERVAL 7 DAY)) * 1000';
		util.logInfo(queryString);
		this.executeQuery(queryString);
		const res = await this.getTradeCounts();
		util.logDebug(`num of trades existing in db ${JSON.stringify(res)}`);
	}
}

const sqlUtil = new SqlUtil();
export default sqlUtil;
