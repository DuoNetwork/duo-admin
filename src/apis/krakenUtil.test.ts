import krakenUtil from './krakenUtil';
import sqlUtil from '../sqlUtil';
const trades: Array<{ [key: string]: string }> = require('../samples/kraken.json');
const apiResponse = JSON.stringify(require('../samples/krakenMsg.json'));

test('parseTrade', () => {
	trades.forEach(async trade => {
		const parsedTrade = krakenUtil.parseTrade(trade);
		expect(parsedTrade).toMatchSnapshot();
	});
});

test('parseApiResponse', async () => {
	sqlUtil.insertSourceData = jest.fn(() => Promise.resolve());
	await krakenUtil.parseApiResponse(apiResponse);
	expect(
		(sqlUtil.insertSourceData as jest.Mock<Promise<void>>).mock.calls[0][0]
	).toMatchSnapshot();
});