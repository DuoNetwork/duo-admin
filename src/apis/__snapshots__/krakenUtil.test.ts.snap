// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseApiResponse 1`] = `
Object {
  "amount": "5.76800000",
  "price": "655.20000",
  "source": "KRAKEN",
  "sourceTimestamp": "1524805049034",
  "tradeId": "1524805049034",
  "tradeType": "buy",
}
`;

exports[`parseTrade 1`] = `
Object {
  "amount": "1.00000000",
  "price": "630.44000",
  "source": "KRAKEN",
  "sourceTimestamp": "1524740175400",
  "tradeId": "1524740175400",
  "tradeType": "sell",
}
`;

exports[`parseTrade 2`] = `
Object {
  "amount": "1.73300000",
  "price": "630.43000",
  "source": "KRAKEN",
  "sourceTimestamp": "1524740175408",
  "tradeId": "1524740175408",
  "tradeType": "sell",
}
`;

exports[`parseTrade 3`] = `
Object {
  "amount": "0.13274374",
  "price": "630.24000",
  "source": "KRAKEN",
  "sourceTimestamp": "1524740175412",
  "tradeId": "1524740175412",
  "tradeType": "sell",
}
`;

exports[`parseTrade 4`] = `
Object {
  "amount": "0.02583083",
  "price": "629.99000",
  "source": "KRAKEN",
  "sourceTimestamp": "1524740181690",
  "tradeId": "1524740181690",
  "tradeType": "buy",
}
`;

exports[`parseTrade 5`] = `
Object {
  "amount": "0.34977855",
  "price": "630.00000",
  "source": "KRAKEN",
  "sourceTimestamp": "1524740181697",
  "tradeId": "1524740181697",
  "tradeType": "buy",
}
`;

exports[`parseTrade 6`] = `
Object {
  "amount": "3.62439062",
  "price": "630.00000",
  "source": "KRAKEN",
  "sourceTimestamp": "1524740181701",
  "tradeId": "1524740181701",
  "tradeType": "buy",
}
`;
